File: src/main/java/org/apache/spark/shuffle/ucx/memory/MemoryPool.java
/*
 * Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
 * See file LICENSE for terms.
 */
package org.apache.spark.shuffle.ucx.memory;

import org.apache.spark.shuffle.UcxShuffleConf;
import org.apache.spark.unsafe.Platform;
import org.openucx.jucx.ucp.UcpContext;
import org.openucx.jucx.ucp.UcpMemory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.nio.ByteBuffer;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.atomic.AtomicInteger;


/**
 * Utility class to reuse and preallocate registered memory to avoid memory allocation
 * and registration during shuffle phase.
 */
public class MemoryPool implements Closeable {
  private static final Logger logger = LoggerFactory.getLogger(MemoryPool.class);

  @Override
  public void close() {
    for (AllocatorStack stack: allocStackMap.values()) {
      stack.close();
      logger.info("Stack of size {}. " +
          "Total requests: {}, total allocations: {}, preAllocations: {}",
        stack.length, stack.totalRequests.get(), stack.totalAlloc.get(), stack.preAllocs.get());
    }
    allocStackMap.clear();
  }

  private class AllocatorStack implements Closeable {
    private final AtomicInteger totalRequests = new AtomicInteger(0);
    private final AtomicInteger totalAlloc = new AtomicInteger(0);
    private final AtomicInteger preAllocs = new AtomicInteger(0);
    private final ConcurrentLinkedDeque<RegisteredMemory> stack = new ConcurrentLinkedDeque<>();
    private final int length;

    private AllocatorStack(int length) {
      this.length = length;
    }

    private RegisteredMemory get() {
      RegisteredMemory result = stack.pollFirst();
      if (result == null) {
        int numBuffers = Math.max(conf.minRegistrationSize() / length, 1);
        preallocate(numBuffers);
        return stack.pollFirst();
      } else {
        result.getRefCount().incrementAndGet();
      }
      totalRequests.incrementAndGet();
      return result;
    }

    private void put(RegisteredMemory registeredMemory) {
      registeredMemory.getRefCount().decrementAndGet();
      stack.addLast(registeredMemory);
    }

    private ByteBuffer malloc(int size) {
      return Platform.allocateDirectBuffer(size);
    }

    private UcpMemory register(ByteBuffer buffer) {
      UcpMemory memory = null;
      if (conf.preregisterMemory()) {
        memory = context.registerMemory(buffer);
      }
      return memory;
    }

    private void preallocate(int numBuffers) {
      // Platform.allocateDirectBuffer supports only 2GB of buffer.
      // Decrease number of buffers if total size of preAllocation > 2GB.
      if ((long)length * (long)numBuffers > Integer.MAX_VALUE) {
        numBuffers = Integer.MAX_VALUE / length;
      }

      ByteBuffer buffer = malloc(numBuffers * length);
      UcpMemory memory = register(buffer);
      AtomicInteger refCount = new AtomicInteger(numBuffers);
      for (int i = 0; i < numBuffers; i++) {
        buffer.position(i * length).limit(i * length + length);
        final ByteBuffer slice = buffer.slice();
        RegisteredMemory registeredMemory = new RegisteredMemory(refCount, memory, slice);
        put(registeredMemory);
      }
      preAllocs.incrementAndGet();
      totalAlloc.incrementAndGet();
    }

    @Override
    public void close() {
      while (!stack.isEmpty()) {
        RegisteredMemory memory = stack.pollFirst();
        if (memory != null) {
          memory.deregisterNativeMemory();
        }
      }
    }
  }

  private final ConcurrentHashMap<Integer, AllocatorStack> allocStackMap =
    new ConcurrentHashMap<>();
  private final UcpContext context;
  private final UcxShuffleConf conf;

  public MemoryPool(UcpContext context, UcxShuffleConf conf) {
    this.context = context;
    this.conf = conf;
  }

  private long roundUpToTheNextPowerOf2(long length) {
    // Round up length to the nearest power of two, or the minimum block size
    if (length < conf.minBufferSize()) {
      length = conf.minBufferSize();
   } else {
      length--;
      length |= length >> 1;
      length |= length >> 2;
      length |= length >> 4;
      length |= length >> 8;
      length |= length >> 16;
      length++;
    }
    return length;
  }

  public RegisteredMemory get(int size) {
    long roundedSize = roundUpToTheNextPowerOf2(size);
    assert roundedSize < Integer.MAX_VALUE && roundedSize > 0;
    AllocatorStack stack =
      allocStackMap.computeIfAbsent((int)roundedSize, AllocatorStack::new);
    RegisteredMemory result = stack.get();
    result.getBuffer().position(0).limit(size);
    return result;
  }

  public void put(RegisteredMemory memory) {
    AllocatorStack allocatorStack = allocStackMap.get(memory.getBuffer().capacity());
    if (allocatorStack != null) {
      allocatorStack.put(memory);
    }
  }

  public void preAlocate() {
    conf.preallocateBuffersMap().forEach((size, numBuffers) -> {
      logger.debug("Pre allocating {} buffers of size {}", numBuffers, size);
      AllocatorStack stack = new AllocatorStack(size);
      allocStackMap.put(size, stack);
      stack.preallocate(numBuffers);
    });
  }

}

File: src/main/scala/org/apache/spark/shuffle/ucx/memory/MemoryPool.scala
/*
* Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
* See file LICENSE for terms.
*/
package org.apache.spark.shuffle.ucx.memory

import java.io.Closeable
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.{ConcurrentHashMap, ConcurrentLinkedDeque}

import org.openucx.jucx.ucp.{UcpContext, UcpMemMapParams, UcpMemory}
import org.openucx.jucx.ucs.UcsConstants
import org.apache.spark.internal.Logging
import org.apache.spark.shuffle.ucx.{MemoryBlock, UcxShuffleConf}
import org.apache.spark.util.Utils

class UcxBounceBufferMemoryBlock(private[ucx] val memory: UcpMemory, private[ucx] val refCount: AtomicInteger,
                                 override val address: Long, override val size: Long)
  extends MemoryBlock(address, size, memory.getMemType == UcsConstants.MEMORY_TYPE.UCS_MEMORY_TYPE_HOST)


/**
 * Base class to implement memory pool
 */
case class MemoryPool(ucxShuffleConf: UcxShuffleConf, ucxContext: UcpContext, memoryType: Int)
  extends Closeable with Logging {

  protected def roundUpToTheNextPowerOf2(size: Long): Long = {
    // Round up length to the nearest power of two
    var length = size
    length -= 1
    length |= length >> 1
    length |= length >> 2
    length |= length >> 4
    length |= length >> 8
    length |= length >> 16
    length += 1
    length
  }

  protected val allocatorMap = new ConcurrentHashMap[Long, AllocatorStack]()

  protected case class AllocatorStack(length: Long, memType: Int) extends Closeable {
    logInfo(s"Allocator stack of memType: $memType and size $length")
    private val stack = new ConcurrentLinkedDeque[UcxBounceBufferMemoryBlock]
    private val numAllocs = new AtomicInteger(0)
    private val memMapParams = new UcpMemMapParams().allocate().setMemoryType(memType).setLength(length)

    private[memory] def get: UcxBounceBufferMemoryBlock = {
      var result = stack.pollFirst()
      if (result == null) {
        numAllocs.incrementAndGet()
        if (length < ucxShuffleConf.minRegistrationSize) {
          preallocate((ucxShuffleConf.minRegistrationSize / length).toInt)
          result = stack.pollFirst()
        } else {
          logInfo(s"Allocating buffer of size $length")
          val memory = ucxContext.memoryMap(memMapParams)
          result = new UcxBounceBufferMemoryBlock(memory, new AtomicInteger(1),
            memory.getAddress, length)
        }
      }
      result
    }

    private[memory] def put(block: UcxBounceBufferMemoryBlock): Unit = {
      stack.add(block)
    }

    private[memory] def preallocate(numBuffers: Int): Unit = {
      logInfo(s"PreAllocating $numBuffers of size $length, " +
        s"totalSize: ${Utils.bytesToString(length * numBuffers) }")
      val memory = ucxContext.memoryMap(
        new UcpMemMapParams().allocate().setMemoryType(memType).setLength(length * numBuffers))
      val refCount = new AtomicInteger(numBuffers)
      var offset = 0L
      (0 until numBuffers).foreach(_ => {
        stack.add(new UcxBounceBufferMemoryBlock(memory, refCount, memory.getAddress + offset, length))
        offset += length
      })
    }

    override def close(): Unit = {
      var numBuffers = 0
      stack.forEach(block => {
        block.refCount.decrementAndGet()
        if (block.memory.getNativeId != null) {
          block.memory.deregister()
        }
        numBuffers += 1
      })
      logInfo(s"Closing $numBuffers buffers of size $length." +
        s"Number of allocations: ${numAllocs.get()}")
      stack.clear()
    }
  }

  override def close(): Unit = {
    allocatorMap.values.forEach(allocator => allocator.close())
    allocatorMap.clear()
  }

  def get(size: Long): MemoryBlock = {
    val roundedSize = roundUpToTheNextPowerOf2(size)
    val allocatorStack = allocatorMap.computeIfAbsent(roundedSize,
      s => AllocatorStack(s, memoryType))
    val result = allocatorStack.get
    new UcxBounceBufferMemoryBlock(result.memory, result.refCount, result.address, size)
  }

  def put(mem: MemoryBlock): Unit = {
    mem match {
      case m: UcxBounceBufferMemoryBlock =>
        val allocatorStack = allocatorMap.get(roundUpToTheNextPowerOf2(mem.size))
        allocatorStack.put(m)
      case _ =>
    }
  }

  def preAllocate(size: Long, numBuffers: Int): Unit = {
    val roundedSize = roundUpToTheNextPowerOf2(size)
    val allocatorStack = allocatorMap.computeIfAbsent(roundedSize,
      s => AllocatorStack(s, memoryType))
    allocatorStack.preallocate(numBuffers)
  }
}

class UcxHostBounceBuffersPool(ucxShuffleConf: UcxShuffleConf, ucxContext: UcpContext)
  extends MemoryPool(ucxShuffleConf, ucxContext, UcsConstants.MEMORY_TYPE.UCS_MEMORY_TYPE_HOST) {
  ucxShuffleConf.preallocateBuffersMap.foreach{
    case (size, numBuffers) => preAllocate(size, numBuffers)
  }
}

class UcxGpuBounceBuffersPool(ucxShuffleConf: UcxShuffleConf, ucxContext: UcpContext)
  extends MemoryPool(ucxShuffleConf, ucxContext, UcsConstants.MEMORY_TYPE.UCS_MEMORY_TYPE_CUDA)
